rfcs/side-effect-lifecycle-scheduling.md

```diff
@@ -0,0 +1,124 @@
+* Consumers are run after producers, readers, and writers. There should only be one consumer of a type of side effect.
+
+### Some Simple Examples
+
+#### Scheduling around command application
+```rs
+fn main() {
+    App::new()
+        .add_system(apply_commands_buffer.consumes("commands"))
+        // if we don't care about when the commands are applied
+        // we can ignore the new api
+        .add_system(create_commands_A)
+        
+        // if we need to schedule a system after commands are applied
+        .add_system(create_commands_A.label("system A").produces("commands"))
+        .add_system(after_commands_A.after("system A").after("commands"))
+}
+```
+
+#### Systems dependent on app state
+```rs
+// this is not an example of how a final api for app state
+// or run criteria is expected to look. Just a way to highlight
+// the relations between the systems that produce those side effects
+
+fn main() {
+    App::new()
+        .add_system(system_that_queues_state_change.produces("state change"))
+        .add_system(system_that_applies_state_queue
+            .consumes("state change")
+            .produces("state")
+        )
+        .add_system(run_criteria
+            .reads("state")
+            .produces("on update state=A")
+        )
+        .add_system(system_A.reads("on update state=A"))
+        .add_system(system_B.reads("on update state=A"))
+}
+```
+
+## Implementation
+
+If we tried to create the relationships between producers, consumers, readers, and writers with system sets and labeling, it would look something like this.
+```rs
+// expressing the subgraph with system sets and labels
+
+fn main() {
+    let SideEffectLabel = Label::new("some side effect");
+    App::new()
+        .add_system_set(SystemSet::new()
+            .label(SideEffectLabel)
+            .add_system_set(SystemSet::new()
+                .label(produces(SideEffectLabel))
+                .before(reads(SideEffectLabel))
+                .before(writes(SideEffectLabel))
+                .before(consumes(SideEffectLabel))
+                .add_system(side_effect_producer)
+            )
+        )
+        // TODO: add consumers, readers, and writer system sets
+}
+```
+
+In practice we can't create these system sets like this as it would be very unergonomic for users. But systems sets are just sugar anyways and don't exist at runtime. We can just apply the correct labels to individual systems.
+
+```rs
+// .produces
+.add_system(side_effect_producer.produces(SideEffectLabel))
+// is equivalent to
+.add_system(side_effect_producer
+    .label(SideEffectLabel)
+    .label(produces(SideEffectLabel))
+    .before(reads(SideEffectLabel))
+    .before(writes(SideEffectLabel))
+    .before(comsumes(SideEffectLabel))
+)
+```
+
+> TODO: add desugaring of readers, writers, and consumers.
+
+## Drawbacks
```

Comment:
IMO the big drawback to this is that it further increases the complexity of the scheduling graph.

This is unavoidable IMO, but users need better tools to inspect and debug this.